hashicorp/waypoint
# ui: improve docker reference parsing

## Why the change?

Closes #2484 and #2495

## What’s the plan?

- [x] Add some failing tests
- [x] Bring in a library to parse these references

## What does it look like?

### Before

| 🌝 | 🌚 |
| --- | --- |
| <img src="https://user-images.githubusercontent.com/34030/137981046-f750f5ef-08dc-4817-8c2e-d6299323832e.png"> | <img src="https://user-images.githubusercontent.com/34030/137980218-22e6bf0e-5109-436c-ab77-86024bb067f5.png"> |

### After

| 🌝 | 🌚 |
| --- | --- |
|  <img width="566" alt="CleanShot 2021-10-22 at 12 33 45@2x" src="https://user-images.githubusercontent.com/34030/138439957-af16f17e-9dde-4cda-a79d-47d1d4bad636.png"> | <img width="566" alt="CleanShot 2021-10-22 at 12 34 03@2x" src="https://user-images.githubusercontent.com/34030/138439965-14be0a14-55ea-4709-8cf9-a012c21f4521.png"> |
| <img width="590" alt="CleanShot 2021-10-22 at 12 34 26@2x" src="https://user-images.githubusercontent.com/34030/138439984-2962a5da-360f-4b87-878e-05e6611234ef.png"> | <img width="590" alt="CleanShot 2021-10-22 at 12 34 11@2x" src="https://user-images.githubusercontent.com/34030/138439995-c069fa79-2bee-442e-8d37-7e02e5290fa7.png"> |

https://user-images.githubusercontent.com/34030/138440695-012b4690-997c-4355-a76c-1f9db4b24c20.mp4

## How do I test it?

1. Pull this branch down `ui/docker-reference`
2. Run the UI dev server
3. Visit [localhost:4200](http://localhost:4200/default/marketing-public/app/wp-matrix/deployment/seq/4)
4. Tweak [the `image` value in the resources factory](https://github.com/hashicorp/waypoint/blob/22a200f43da02a53b24b8006df93c20bcdd222cb/ui/mirage/factories/resource.ts#L60)
5. Verify these values show up correctly in the UI

Some examples to test with:

```
alpine
alpine@sha256:aaaaf56b44807c64d294e6c8059b479f35350b454492398225034174808d1726
registry:2
jpetazzo/pxe
jpetazzo/pxe:latest
quay.io/signalfuse/zookeeper
index.docker.io/library/ubuntu:latest
internal.mycorp.com:5000/revealjs
internal.mycorp.com:5000/revealjs:3.4.5-3
localhost:5000/mart/mass:latest
localhost:5000/mass
```
